src/popup/App.tsx

```diff
@@ -20,127 +20,145 @@ import emptyInbox from './assets/empty_inbox.svg';
 import './style.css';
 import { MergeRequestsDetails } from '../background/types';
 import { getHumanReadableDate } from './helpers';
+import { getMergeRequestList, MergeRequestSendMessageReply } from '../libs/mergeRequestDownloader';
+
+type AppStatus = 'idle' | 'loading' | 'success' | 'error';
 
 const App = () => {
-    const fetchingContentState = <Text>Fetching content...</Text>;
-    const [mrList, updateList] = useState(fetchingContentState);
-    const [mrToReview, setMrToReview] = useState('0');
-    const [mrReviewed, setMrReviewed] = useState('0');
-    const [mrRatio, setMrRatio] = useState(100);
-    const [lastUpdateDateUnix, setLastUpdateDateUnix] = useState(0);
+    const [appStatus, setAppStatus] = useState<AppStatus>('idle');
+    const [errorMessage, setErrorMessage] = useState('');
+    const [mrData, setMrData] = useState<MergeRequestSendMessageReply>({
+        mrAssigned: [],
+        mrToReview: 0,
+        mrGiven: [],
+        mrReviewed: 0,
+        lastUpdateDateUnix: Date.now()
+    });
 
-    const [tabSelected, setTabSelected] = useState('0');
+    const [currentTab, setCurrentTab] = useState(0);
 
     const openSettings = useCallback(() => browser.runtime.openOptionsPage(), []);
 
-    let type = 'getMRs';
-
-    interface SendMsgResponse {
-        mrAssigned: MergeRequestsDetails[];
-        mrToReview: number;
-        mrGiven: MergeRequestsDetails[];
-        mrReviewed: number;
-        lastUpdateDateUnix: number;
-    }
-    interface SendMsgError {
-        error: string;
-    }
-    const sendMsg = useCallback((event: any) => {
-        if (event && event.target.dataset.key) {
-            setTabSelected(event.target.dataset.id);
-            type = event.target.dataset.key;
-        }
-        browser.runtime
-            .sendMessage({ type: 'getMRs' })
-            .then((response: SendMsgResponse | SendMsgError) => {
+    // useEffect vs useCallback
+    // https://medium.com/@infinitypaul/reactjs-useeffect-usecallback-simplified-91e69fb0e7a3
+    const fetchData = useCallback(() => {
+        setAppStatus('loading');
+        getMergeRequestList()
+            .then((response) => {
                 if (!response) {
-                    // This could infinite loop, so let's see for a correct retry
-                    console.log('Retrying');
-                    return setTimeout(sendMsg(type), 1000);
+                    setAppStatus('error');
+                    setErrorMessage('Something went wrong');
+                    return;
                 }
 
-                if ('error' in response) {
-                    return updateList(
-                        <Flash m={2} scheme="red">
-                            {response.error}
-                        </Flash>
-                    );
+                setMrData(response);
```

Comment:
This should go after all errors check, otherwise it set the data with the error content.

Reply:
true

Reply:
fixed